onnxruntime/core/providers/nuphar/mti_x86/nn/pool_ops.cc

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/nuphar/mti_x86/nn/pool_ops.h"
+
+#include "core/codegen/mti/mti_tvm_utils.h"
+#include "core/mlas/inc/mlas.h"
+#include "core/providers/cpu/nn/pool_attributes.h"
+#include <topi/detail/extern.h>
+
+namespace onnxruntime {
+namespace nuphar {
+
+TVM_REGISTER_GLOBAL("tvm.contrib.onnxruntime.pool_f32")
+    .set_body([](tvm::TVMArgs args, tvm::TVMRetValue* /*ret*/) {
+      // input
+      DLTensor* X = args[0];
+      DCHECK(tvm::runtime::TypeMatch(X->dtype, kDLFloat, 32));
+      // output
+      DLTensor* Y = args[1];
+      DCHECK(tvm::runtime::TypeMatch(Y->dtype, kDLFloat, 32));
+
+      // enum is not an integral type
+      int k = args[2];
+      MLAS_POOLING_KIND kind = static_cast<MLAS_POOLING_KIND>(k);
+
+      int num_args = args.size();
+      DCHECK(num_args > 3);
+      int arg_idx = 3;
+
+      auto extract_values_fn = [&]() {
+        std::vector<int64_t> vec;
+
+        DCHECK(arg_idx < num_args);
+        int64_t num_vec = args[arg_idx++];
+        for (int i = 0; i < num_vec; i++, arg_idx++) {
+          DCHECK(arg_idx < num_args);
+          int64_t v = args[arg_idx];
+          vec.push_back(v);
+        }
+        return vec;
+      };
+
+      std::vector<int64_t> kernel_shape = extract_values_fn();
+      std::vector<int64_t> padding = extract_values_fn();
+      std::vector<int64_t> strides = extract_values_fn();
+
+      MlasPool(kind,
+               /*num_pooling_dims*/ X->ndim - 2,
```

Comment:
>2 [](start = 46, length = 1)

if MlasPool currently only supports 2D, please check if X->ndim == 4. Otherwise, try use X->ndim - kernel_shape.size() to be more general

Reply:
MlasPool supports 1D/2D/3D. ```X->ndim - 2``` == ```kernel_shape.size()```. Probably I should change it to kernel_shape.size(). It's more straightforward.